### ISSUE TITLE
Repeated calls to subscripts seem to blow up the stack

### ISSUE THREAD

**Issue by kyouko-taiga** [hylo-lang/hylo#1474]
Compile this program with `-O`:

```swift
fun primes_until(_ n: Int) -> Int {
  var s = Array<Bool>(count: n, initialized_with: fun(_ i) { true })
  &s[0] = false
  &s[1] = false
  var i = 0
  while true {
    var j = i * i
    if j >= n {
      break
    } else if s[i] {
      while j < n {
        &s[j] = false
        &j += i
      }
    }
    &i += 1
  }

  var r = 0
  &i = 0
  while i < n {
    if s[i] { &r += 1 }
    &i += 1
  }
  return r
}

public fun main() {
  let a = primes_until(1000000)
  print(a)
}
```

Execution consistently crashes in the second loop. On summary inspection, it seems that we're running out of stack. The program behaves as expected for smaller loop bounds.

**Comment by kyouko-taiga** [comment]
Actually the issue doesn't seem to be about repeated calls. This program runs as expected (i.e., it never terminates):
```swift
public fun main() {
  let a = Array<Int>(count: 10, initialized_with: fun (_ i) { i.copy() })
  var i = 0
  while true {
    var x = 1
    if a[i] != 0 { &x = 0 }
    &i = (i + 1) % 10
  }
}
```

**Comment by Metalymph** [comment]
I've seen that `if s[i] { &r += 1 }` is equivalent to `if s[i] { }` in terms of executions (always 522.7K over stack limit). So (apart that `if s[i] { }` should be lowered as a no-op zero allocation stmt, right?) it seems like each temporary projection of `s[i]` is kept into the stack instead of being overwritten.

**Comment by kyouko-taiga** [comment]
Yes, that matches my suspicion, though I haven't been able to confirm it yet. I should look at the generated LLVM IR.

If that's the case then our reliance on LLVM's coroutines to compile subscript is a liability and we should seriously consider an alternative. @dabrahams has figured out an efficient approach but implementing it is quite challenging.

**Comment by dabrahams** [comment]
If swift can do it, we should be able to also, no?

**Comment by kyouko-taiga** [comment]
No because Swift has a much simpler problem to solve. It doesn't need to perform allocation between the creation of a projection and its last use, because subscripts can only be used as subexpressions.
